src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,205 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class MakeFieldReadonlyDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public MakeFieldReadonlyDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.ProjectAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext =>
```

Comment:
❕ Refactor this callback to be a method (not local function) named `HandleCompilationStart`

Reply:
I disagree with this suggestion - all analyzer actions and nested actions **should** be registered in a single method to make it easier to understand the overall analysis logic. We follow this across all analyzers in the repos we own and was also the primary reason that the design team decided on analyzer actions being delegates during original analyzer API design.

I have added more comments and refactored out logic within the registered actions into separate helper methods to improve readability.

Reply:
> I disagree with this suggestion - all analyzer actions and nested actions should be registered in a single method to make it easier to understand the overall analysis logic.

@mavasani  Your reasoning seems fine.  *Except* for the fact that you're going against the patterns and practices the IDE uses.  I'm not a fan of this as recently several PR have been coming in that do this sort of thing.  If you want to change the patterns here, please discuss it with the relevant stakeholders, and come to consensus on these sorts of things.  Then, we can decide if we like it, and if so *uniformly* move all the code to be consistent *or* we can decide "no, we like it as is".

Right now, these PRs are increasing inconsistency in the IDE codebase.  And that's not good.  If you believe this approach is better, it can be discussed.  But it should only go forward with team acceptance, and with a plan to ensure we don't have total inconsistency across the features.

> We follow this across all analyzers in the repos we own and was also the primary reason that the design team decided on analyzer actions being delegates during original analyzer API design.

I do not recall that part of the discussion at all.  The primary reason i remember being stated was so that people could easily register callbacks in strongly typed fashion.  i.e. i would ask to be called back for some event, and i would be given the context that i could only do the right thing with.  That was better than a whole host of events and callbacks on a single object that you would have to know how they all worked together.

Also, a common pattern we showed when during this time was simply:  ```x.RegisterXXXAction(() => DoSomething(), Kinds);```  It was never a core outcome that the expected or desired way was to put all logic in the lambda.  That was always up to the final user depending on if they felt such an approach was more pleasant or not.